src/refs/immutable.rs

```diff
@@ -201,7 +211,7 @@ where
     /// }
     /// ```
     #[inline]
-    pub unsafe fn into_split(r: Ref<'a, T>) -> (&'a T, AtomicBorrow<'a>) {
+    pub unsafe fn into_split(r: Ref<'a, T>) -> (NonNull<T>, AtomicBorrow<'a>) {
         (r.value, r.borrow)
     }
```

Comment:
This function can be safe now since it is unsafe to read from the `NonNull<T>`

Reply:
oh and I assume there is a similar situation for `RefMut`, although users of that need to be careful since `NonNull<T>` is covariant over `T`

Reply:
I think it makes sense to leave these `unsafe` because of the extra soundness invariants associated with it that are not documented with `NonNull<T>`, i.e. that it is only valid to dereference the `NonNull` if the atomic guard has not been dropped.

---

Good point with the `NonNull` being covariant. As an experiment, I tested out a `NonNullMut<T>` wrapper type that is invariant over `T`, which seems useful in this context. This allows the user to store the `NonNullMut<T>` in whatever guard type they want along with the `AtomicBorrow`, without having to also add and phantom data to make it invariant.

Do you think there are any problems with that approach?

Reply:
I don't see any particular issues with that other than just the implicit complexity overhead of introducing a new type to work with. 

Ideally, if `MaybeDangling<T>` existed in `std`, then `MaybeDangling<&'a mut T>` might work well.

----

> I think it makes sense to leave these unsafe because of the extra soundness invariants associated with it that are not documented with `NonNull<T>`

The thing is, a `NonNull` considered by itself doesn't give any guarantee that it is ever valid to dereference. The user dereferencing a `NonNull` must always be aware of its origins to ensure proper alignment, that it is pointing to a valid allocation with a valid instance of `T`, there are no aliasing conflicts, the pointer has the correct provenance for the access being performed, etc. So rather than the API here applying additional invariants that must be met to soundly deference the `NonNull`, instead it is providing the guarantee that the `NonNull` will meet all these existing requirements if the atomic guard has not been dropped.

Reply:
> So rather than the API here applying additional invariants that must be met to soundly deference the NonNull, instead it is providing the guarantee that the NonNull will meet all these existing requirements if the atomic guard has not been dropped.

Ah, yeah, that perspective makes sense to me. 👍️ Updated it to be safe.

---

Regarding the `NonNullMut` wrapper, I'm really on the fence whether or not to document the extra invariant, keep the wrapper type, or use the [`maybe-dangling`](https://docs.rs/maybe-dangling/latest/maybe_dangling/struct.MaybeDangling.html) crate.

@zakarumych do you have an opinion on this?

Reply:
I would keep simple `NonNull` and add `PhantomData<&mut T>` to `RefMut` to make it invariant over `T`.